File: python/Day02-Exer02.py
# Vùng chứa nhiều nước nhất - Bạn được cung cấp một mảng số nguyên có chiều cao, chiều dài, n. 
# Có n đường thẳng đứng được vẽ sao cho hai điểm cuối của đường thứ i là (i, 0) và (i, chiều cao[i]). 
# Tìm hai đường cùng với trục x tạo thành một thùng chứa, sao cho thùng chứa nhiều nước nhất (độ sâu giữa 
# các thùng chứa không đổi). Trả về diện tích (mà 2 đường và trục X tạo thành) của thùng chứa có thể chứa lượng nước tối đa. 
# Lưu ý rằng bạn không được nghiêng thùng chứa.

#! By me:
def max_area(array):
    if len(array) < 2:
        return 0
    max_area = 0
    area = 0
    for i in range(len(array)):
        for j in range(len(array)):
            if array[i] >= array[j]:
                area = array[j] * (i - j)
            else:
                area = array[i] * (i - j)
            if area > max_area:
                max_area = area
    return max_area

#! Bruteforce
def max_area(array):
    max_area = 0
    for i in range(len(array)-1):
        for j in range(i+1,len(array)):
            area = min(array[i],array[j])*(j-i)
            max_area = max(max_area,area)
    return max_area

#! Optimized
def max_area_optimum(array):
    left = 0
    right = len(array)-1
    max_area=0
    while(left<right):
        area = min(array[left], array[right]) * (right - left)
        max_area = max(area,max_area)
        if array[left]<array[right]:
            left +=1
        else:
            right -= 1
    return max_area

File: javascript/Day02-Exer02.js
// Vùng chứa nhiều nước nhất - Bạn được cung cấp một mảng số nguyên có chiều cao, chiều dài, n.
// Có n đường thẳng đứng được vẽ sao cho hai điểm cuối của đường thứ i là (i, 0) và (i, chiều cao[i]).
// Tìm hai đường cùng với trục x tạo thành một thùng chứa, sao cho thùng chứa nhiều nước nhất (độ sâu giữa
// các thùng chứa không đổi). Trả về diện tích (mà 2 đường và trục X tạo thành) của thùng chứa có thể chứa lượng nước tối đa.
// Lưu ý rằng bạn không được nghiêng thùng chứa.

//! By me:

const maxArea = function (array) {
  if (array.length == 0 || array.length == 1) return 0;
  let maxArea = 0;
  let area = 0;

  for (let i = 0; i < array.length; i++) {
    for (let j = 0; j < array.length; j++) {
      if (array[i] >= array[j]) {
        area = Math.abs((j - i) * array[j]);
      }
      if (area > maxArea) {
        maxArea = area;
      }
    }
  }
  return maxArea;
};

const a = [2, 6, 7, 5, 2];
console.log(maxArea(a));

//! Bruteforce
// Time complexity: O(n^2) - Space complexity: O(1)
const maxAreaBruteForce = function (array) {
  //[3,7,5,6,8,4]
  let area = 0;
  for (let i = 0; i < array.length - 1; i++) {
    for (let j = i + 1; j < array.length; j++) {
      const height = Math.min(array[i], array[j]);
      const width = j - i;
      area = Math.max(area, height * width);
    }
  }
  return area;
};

//! Optimized
// Time complexity: O(n) - Space complexity: O(1)
const maxAreaOP = function (array){
  let area =0;
  let i=0;
  let j=array.length-1;
  while(i<j){
      let height = Math.min(array[i],array[j]);
      let newArea = height*(j-i);
      area = Math.max(area,newArea);
      if(array[i]<array[j]) i++;
      else j--;
  }
  return area;
}
